### PR TITLE
Implement a Markdown AST Parser

### LINKED ISSUE #45: Ensure that the paragraph checks disregard markdown syntax
Currently there are some tests to check paragraph counting with Markdown. However, we should ensure that this works correctly and add more tests to the `test_fragments` suite.

### LINKED ISSUE #81: Consider the adoption of a general-purpose Markdown parser
There are some current issues, like #45 and #73, that will require us to revisit the way in which we parse the Markdown content. Currently, this parsing happens through the use of regular expressions. In the future, it would be much better to use a Python-based Markdown parser that produces a parse tree. We can then traverse this parse tree and check that the correct aspects of Markdown are evident. Ultimately, this approach is the most likely to support flexible and correct checking of Markdown files.

### PR DESCRIPTION
**Pull request by huangs1**
This PR fixes #81 and fixes #45. Team 3 implemented a replacement for the current function `get_paragraphs` in `fragments.py`. We've removed the function `is_paragraph` in `fragments.py` since the old  `get_paragraphs` function was the only one that used it. Also, we have added the function `is_single_subnode` that indicates what shouldn't be included as a paragraph.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -6,14 +6,15 @@ name = "pypi"
...
+commonmark = "*"
 
 [dev-packages]
 pylint = "*"
 black = "*"
 pytest = "*"
 six = "*"
 codecov = "*"
-"flake8" = "*"
+flake8 = "*"
```
Comment (Michionlion) [inline Pipfile]: We should ask @gkapfham to confirm that removing the `"` surrounding `flake8` here will have no negative consequences, just to be sure.

File: gator/fragments.py [hunk 2]
```diff
@@ -1,64 +1,59 @@
...
-        return False
-    # there may be other edge cases that should be added here in the
-    # future -- what other structures look like paragraphs but should
-    # not be?
-    # if nothing has returned by now, the candidate must be a paragraph
-    return True
+def is_single_subnode(subnode):
+    """Determines if a node is single"""
+    return (
+        # Excludes the following subnode types
```
Comment (Michionlion) [inline gator/fragments.py]: This documentation comment:
* Does not belong here (documentation should be placed in the docstring above).
* Is not fully accurate (what are they excluded from?)

File: gator/fragments.py [hunk 3]
```diff
@@ -1,64 +1,59 @@
...
-    paragraphs = pattern.findall(contents)
-    # disregard all of the section headers in markdown
-    matching_paragraphs = []
-    # iterate through all potential paragraphs and gather
-    # those that match the standard for legitimacy
-    for paragraph in paragraphs:
-        if is_paragraph(paragraph) is True:
-            matching_paragraphs.append(paragraph)
-    return matching_paragraphs
+    # Calls the CommonMark parser
```
Comment (Michionlion) [inline gator/fragments.py]: I'm not sure this comment is necessary. The review team (@finneyj2, @durcij) should have a discussion on whether inline documentation comments should be used or not. If they should, perhaps more should be added. Please work with the implementation team to determine what those comments should be, if you decide they are needed!

File: gator/fragments.py [hunk 4]
```diff
@@ -89,16 +84,14 @@ def is_blank_line(line):
...
+    matching_paragraphs = get_paragraphs(contents)
     return len(matching_paragraphs)
 
 
 def count_words(contents):
     """Counts the minimum number of words across all paragraphs in writing"""
     # retrieve all of the paragraphs in the contents
-    replace_blank_inputs = False
-    paragraphs = get_paragraphs(contents, replace_blank_inputs)
+    paragraphs = get_paragraphs(contents)
```
Comment (Michionlion) [inline gator/fragments.py]: @gkapfham what was the use of `replace_blank_inputs` here? It was not clear from the source code exactly what mattered, and although I think now that we use an AST to parse the markdown it should not matter, I would like to confirm that with you.

File: tests/test_fragments.py [hunk 5]
```diff
@@ -15,24 +15,16 @@
         ("hello world!!%^(@after)writing a lot\n", 1),
         ("hello world!!%^(@after)writing a lot\n\n", 1),
         ("", 0),
-        ("", 0),
-        (" ", 0),
```
Comment (Michionlion) [inline tests/test_fragments.py]: @gkapfham these tests are original code from before we started on this branch. Is there any use to duplicate test cases? Perhaps to catch flaky tests?

File: gator/fragments.py [hunk 6]
```diff
@@ -1,64 +1,59 @@
...
-        if is_paragraph(paragraph) is True:
-            matching_paragraphs.append(paragraph)
-    return matching_paragraphs
+    # Calls the CommonMark parser
+    ast = commonmark.Parser().parse(contents)
+    mode_looking = True
+    paragraph_list = []
+    paragraph_content = ""
+    counter = 0
+    for subnode, enter in ast.walker():
```
Comment (finneyj2) [inline gator/fragments.py]: It could be useful to add some comments in this For Loop to increase clarity

File: gator/fragments.py [hunk 7]
```diff
@@ -1,64 +1,65 @@
...
-        if is_paragraph(paragraph) is True:
-            matching_paragraphs.append(paragraph)
-    return matching_paragraphs
+    ast = commonmark.Parser().parse(contents)
+    mode_looking = True
+    paragraph_list = []
+    paragraph_content = ""
+    counter = 0
+    for subnode, enter in ast.walker():
+        """Iterates through the file to find the paragraphs and add them to the array"""
```
Comment (Michionlion) [inline gator/fragments.py]: Docstrings only work at the beginning of a function -- for inline code comments, one should use `#`. @finneyj2, @durcij, can you fix this so our PR passes Travis?

Comment (Michionlion) [reply gator/fragments.py]: @durcij, @finneyj2, there are still other places where a docstring must be replaced. Remember, docstrings should only be used at the beginning of functions. All other documentation should be in comments.

Comment (durcij) [reply gator/fragments.py]: Forgot that one was there, it should be fixed now.  Thanks for the heads-up!

File: gator/fragments.py [hunk 8]
```diff
@@ -1,64 +1,65 @@
...
+            # Checks to see if it has found the closing of the paragraph subnode
+            if counter == 2 and subnode.t == "paragraph" and not enter:
+                # Adds the content of the paragraph to the array
+                paragraph_list.append(paragraph_content)
+                # Starts a search for a new paragraph
+                mode_looking = True
+            if subnode.literal is not None:
+                paragraph_content += subnode.literal
+
+        # Looks for an opening and a closing node
```
Comment (Michionlion) [inline gator/fragments.py]: This documentation comment is misleading -- this code does not look for an opening and closing node, it keeps track of the depth inside the tree we are currently at.

**Comment by Michionlion** [comment]
We should use the [`is_container`](https://github.com/rtfd/CommonMark-py/blob/master/commonmark/node.py#L6) method from the CommonMark-py implementation to simplify our implementation.

File: tests/test_fragments.py [hunk 9]
```diff
@@ -5,6 +5,32 @@
...
 
+@pytest.mark.parametrize(
+    "writing_string,expected_count",
+    [
+    ("Paragraph1\n - first item\n - second item\n\nParagraph2", 2),
+    ("Paragraph1\n 1. item one\n 2. item two\n\nParagraph2", 2),
+    (" - first item\n - second item\n", 0),
+    ]
+)
+def test_listsbreak_paragraphs(writing_string, expected_count):
```
Comment (Michionlion) [inline tests/test_fragments.py]: Hi @lylet-AC, instead of adding a new test method, could you add your cases to the correct existing test method?

**Comment by gkapfham** [comment]
Hello @huangs1, can you please look into why the checks are not passing for this build. I would like to soon add this feature to GatorGrader by merging in this PR. Also, what is the status of our recent conversation about implementing this with a separate feature/command-line argument exclusively for checks about Markdown files?

**Comment by huangs1** [comment]
Hello @gkapfham, @Michionlion has fixed the formatting error that was made previously. The build is now building successfully. As for implementing a separate command-line argument for checks in Markdown files, we are currently in the process implementing a check for code blocks and we plan on adding more checks in the future.

File: tests/test_fragments.py (hunk 5 above)
Comment (gkapfham) [reply tests/test_fragments.py]: Hi @Michionlion, I don't think that those two duplicate tests for `""` are both needed.

File: gator/fragments.py (hunk 4 above)
Comment (gkapfham) [reply gator/fragments.py]: I don't see an obvious need for the `replace_blank_inputs` variable, @Michionlion. Did you ever discover a reason for this variable? It seems that your new approach obviates the need for it, right?

Comment (Michionlion) [reply gator/fragments.py]: That is correct, @gkapfham! I'll go with this implementation then.

**Comment by Michionlion** [comment]
> I can see that the build is passing and that your new approach seems to achieve the goals that we set for it. While I'm glad to see this approved, I'm marking this as a request for changes because I'm wondering why the commits for several individuals, like @ohnoanarrow, are appearing with a default GitHub icon instead of the correct photo? Will these more "anonymous" commits appear that way in the final commit log? I would like to ensure that @ohnoanarrow and others get full credit for the commits that they made to the repository.

@ohnoanarrow Can you update your local git email so that this will not happen in the future? You should use the email you registered with Github. Unfortunately, @gkapfham, it's unlikely that we can easily change the author of these past commits.

**Comment by Michionlion** [comment]
I'm merging this because a new bug has just been reported that is fixed by this PR, involving counting code blocks with new lines in them as paragraphs. If there are any concerns about this PR or changes involved, let me know!

**Pull request merged**